src/main/gui/main-window.ts

```diff
@@ -190,38 +182,78 @@ const registerEventHandlerPostSetup = (
     ipcMain.on('update-has-unsaved-changes', (_, value) => {
         hasUnsavedChanges = value;
     });
+
     let forceExit = false;
     ipcMain.on('exit-after-save', () => {
         forceExit = true;
         mainWindow.close();
     });
 
+    ipcMain.on('reboot-after-save', () => {
+        app.relaunch();
+        app.exit();
+    });
+
+    const handleUnsavedChanges = (
+        event: Electron.Event,
+        onSave: () => void,
+        onDontSave?: () => void
+    ) => {
+        const choice = dialog.showMessageBoxSync(mainWindow, {
+            type: 'question',
+            title: 'Unsaved changes',
+            message: 'The current chain has unsaved changes.',
+            buttons: ['&Save', "Do&n't Save", 'Cancel'],
+            defaultId: 0,
+            cancelId: 2,
+            noLink: true,
+            normalizeAccessKeys: true,
+        });
+        if (choice === 1) {
+            // Don't save
+            if (onDontSave) {
+                onDontSave(); // Restart the application
+            }
+        } else if (choice === 2) {
+            // Cancel
+            event.preventDefault();
+        } else {
+            // Save
+            event.preventDefault();
+            onSave();
+        }
+    };
+
+    ipcMain.handle('relaunch-application', (event) => {
+        if (backend.owned) {
+            backend.tryKill();
+        }
```

Comment:
So when the user selects "Cancel", the backend will be killed the UI will stay open?

Reply:
Oh, thank you for pointing this out.